### PR TITLE
Provides the ability to customize the apikey generation and signing.

### PR DESCRIPTION
**Pull request by ShalkiWenushika**
This PR provides the ability to customize the API key generation and signing.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java:2]: Change the license header to 2020

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java [hunk 2]
```diff
@@ -1,136 +1,39 @@
...
+ */
+public interface ApiKeyGenerator {
+    /**
+     * Generates and signs the apikey, which can be used for the authorization of APIs deployed at gateways
+     * @param jwtTokenInfoDTO
+     * @return the signed apikey as a string
+     * @throws org.wso2.carbon.apimgt.api.APIManagementException
+     */
+    String generateToken(JwtTokenInfoDTO jwtTokenInfoDTO) throws APIManagementException;
+}
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java:39]: add new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    private static String encode(byte[] stringToBeEncoded) throws APIManagementException {
+        try {
+            return java.util.Base64.getUrlEncoder().encodeToString(stringToBeEncoded);
+        } catch (Exception e) {
+            throw new APIManagementException("Error while encoding the Api Key ",e);
+        }
+    }
+
+}
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java:175]: add new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+import org.wso2.carbon.base.api.ServerConfigurationService;
+import org.wso2.carbon.core.util.KeyStoreManager;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.nio.charset.Charset;
+import java.security.*;
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java:35]: remove wild card import

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        long expireIn;
+        if (jwtTokenInfoDTO.getExpirationTime() == -1 ||
+                jwtTokenInfoDTO.getExpirationTime() > (Integer.MAX_VALUE-currentTime)) {
+            expireIn = -1;
+        } else {
+            expireIn = currentTime + jwtTokenInfoDTO.getExpirationTime();
+        }
+        String issuerIdentifier = OAuthServerConfiguration.getInstance().getOpenIDConnectIDTokenIssuerIdentifier();
+        JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder();
+        jwtClaimsSetBuilder.claim("sub", APIUtil.getUserNameWithTenantSuffix(jwtTokenInfoDTO.getEndUserName()));
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/DefaultApiKeyGenerator.java:83]: use the constants declared in JwtTokenConstants

**Comment by dushaniw** [comment]
@ShalkiWenushika Please address the merge conflicts

**Review by mushthaq33 — APPROVED** [review]

**Pull request merged**